File: 小红的魔法树探险.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl '\n'
#define int long long
const int mod = 1e9 + 7;

int ksm(int a, int b, int mod) {
    a %= mod;
    int ans = 1;
    while (b) {
        if (b & 1) {
            ans = (ans * a) % mod;
        }
        a = (a * a) % mod;
        b >>= 1;
    }
    return ans;
}

void solve() {
    int n, total_exp = 0;
    cin >> n;
    vector<int> g[n + 1], in(n + 1), p(n + 1), vis(n + 1);
    for (int i = 1; i < n; ++i) {
        int u, v;
        cin >> u >> v;
        g[u].push_back(v);
        g[v].push_back(u);
        in[u]++;
        in[v]++;
    }
    p[1] = vis[1] = 1;
    queue<int> q;
    q.emplace(1);
    while (q.size()) {
        int u = q.front(), nowp = p[q.front()];
        q.pop();
        total_exp = (total_exp + nowp) % mod;
        int inv_in = ksm(in[u], mod - 2, mod);
        int nextp = nowp * inv_in % mod;
        for (int v : g[u]) {
            if (not vis[v]) {
                vis[v] = 1;
                p[v] = nextp;
                q.emplace(v);
            }
        }
    }
    cout << total_exp << endl;
}

signed main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    cout << fixed << setprecision(15);
    solve();
    return 0;
}

File: 小红的魔法树探险.py
import sys
from collections import deque
input = lambda: sys.stdin.readline().strip()
mod = 10 ** 9 + 7 # 998244353

n = int(input())
g, IN = [[] for _ in range(n + 1)], [0 for _ in range(n + 1)]
for _ in range(n - 1):
    u, v = map(int, input().split())
    g[u].append(v)
    g[v].append(u)
    IN[u] += 1
    IN[v] += 1
q, p = deque(), [0 for _ in range(n + 1)]
p[1] = 1
vis = [0 for _ in range(n + 1)]
vis[1] = 1
q.append(1)
total_exp = 0
while q:
    u = q.popleft()
    nowp = p[u]
    total_exp = (total_exp + nowp) % mod
    inv_IN = pow(IN[u], mod - 2, mod)
    nextp = nowp * inv_IN % mod
    for v in g[u]:
        if not vis[v]:
            vis[v] = 1
            p[v] = nextp
            q.append(v)
print(total_exp)

